homeassistant/components/sensor/iperf3.py

```diff
@@ -0,0 +1,168 @@
+
+    for sensor in dev:
+        hass.services.register(DOMAIN, sensor.service_name, update)
+
+
+class Iperf3Sensor(Entity):
+    """A Iperf3 sensor implementation."""
+
+    def __init__(self, server, port, duration, sensor_type):
+        """Initialize the sensor."""
+        self._attrs = {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+        }
+        self._name = \
+            "{} {}".format(SENSOR_TYPES[sensor_type][0], server)
+        self._state = None
+        self._sensor_type = sensor_type
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+        self._port = port
+        self._server = server
+        self._duration = duration
+        self.result = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def service_name(self):
+        """Return the service name of the sensor."""
+        return slugify("{} {}".format('update_iperf3', self._server))
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self.result is not None:
+            self._attrs[ATTR_ATTRIBUTION] = CONF_ATTRIBUTION
+            self._attrs[ATTR_PROTOCOL] = self.result.protocol
+            self._attrs[ATTR_REMOTE_HOST] = self.result.remote_host
+            self._attrs[ATTR_REMOTE_PORT] = self.result.remote_port
+            self._attrs[ATTR_VERSION] = self.result.version
+        return self._attrs
+
+    def update(self):
+        """Get the latest data and update the states."""
+        import iperf3
+        client = iperf3.Client()
+        client.duration = self._duration
+        client.server_hostname = self._server
+        client.port = self._port
+        client.verbose = False
+
+        # when testing download bandwith, reverse must be True
+        if self._sensor_type == 'download':
+            client.reverse = True
+
+        try:
+            self.result = client.run()
+        except (OSError, AttributeError) as error:
+            self.result = None
+            _LOGGER.error("Iperf3 sensor error: %s", error)
+            return
+
+        if self.result is not None and \
+           hasattr(self.result, 'error') and \
+           self.result.error is not None:
+            _LOGGER.error("Iperf3 sensor error: %s", self.result.error)
+            self.result = None
+            return
+
+        if self._sensor_type == 'download':
+            self._state = round(self.result.received_Mbps, 2)
+
+        elif self._sensor_type == 'upload':
+            self._state = round(self.result.sent_Mbps, 2)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
```

Comment:
We should not restore state from the database since it's queried right on startup.